packages/test/test-service-load/src/gcDataStore.ts

```diff
@@ -1,101 +1,155 @@
+/**
+ * Root data object for the stress tests that does the following when asked to run:
+ * - It sends ops at a regular interval. The interval is defined by the config passed to the run method.
+ * - After every few ops, it does an activity. Example activities:
+ *     - Create a child data store, reference it and asks it to run.
+ *     - Ask a child data store to stop running and unreferenced it.
+ */
+export class RootDataObject extends BaseDataObject implements IGCDataStore {
     public static get type(): string {
-        return "DataObjectParent";
+        return "RootDataObject";
     }
-    protected async hasInitialized(): Promise<void> {
-        await super.hasInitialized();
 
-        this.child = await dataObjectWithCounterFactory.createInstance(this.context.containerRuntime);
-    }
+    private shouldRun: boolean = false;
+
+    private readonly uniqueChildKey = `${uuid()}-child`;
+    private child: IGCDataStore | undefined;
 
     public async run(config: IRunConfig) {
-        const childP = this.child?.run(config);
-        const parentP = super.run(config);
-        const allP = await Promise.all([childP, parentP]);
-        return allP.every((done) => done);
+        this.shouldRun = true;
+        const delayBetweenOpsMs = 60 * 1000 / config.testConfig.opRatePerMin;
+        const clientSendCount = config.testConfig.totalSendCount / config.testConfig.numClients;
+        let localSendCount = 0;
+
+        while (this.shouldRun && localSendCount < clientSendCount && !this.runtime.disposed) {
+            if (localSendCount % 3 === 0) {
+                console.log(
+                    `########## GC DATA STORE [${this.runtime.clientId}]: ${localSendCount} / ${this.counter.value}`);
+            }
+
+            // After every 3 ops, perform an activity.
+            if (localSendCount % 3 === 0) {
+                await this.preformActivity(config);
+            }
+
+            this.counter.increment(1);
+            localSendCount++;
+            await delay(delayBetweenOpsMs);
+        }
+        return !this.runtime.disposed;
+    }
+
+    public stop() {
+        this.shouldRun = false;
+    }
+
+    /**
+     * Perform the following activity:
+     * - Ask the current child data store to stop running and unreferenced it.
+     * - Create a child data store, reference it and asks it to run.
+     */
+    private async preformActivity(config: IRunConfig) {
+        this.child?.stop();
+        this.child = await dataObjectType1Factory.createInstance(this.context.containerRuntime);
+        // This will unreference the previous child and reference the new one.
+        this.root.set(this.uniqueChildKey, this.child.handle);
+
+        // Set up the child to send ops 10 times faster than this data store.
+        const childConfig: IRunConfig = { ...config };
+        childConfig.testConfig.opRatePerMin = config.testConfig.opRatePerMin * 10;
+
+        this.child.run(childConfig).then((done: boolean) => {
```

Comment:
Remember, you gotta await this if you want the error to propagate.

Also, this error thrown from the `.then` will get caught by the `.catch`, JFYI.  you can also do 

```
.then((done: boolean) => { throw ChildDisposedError(); }, (error) => { throw GenericError(); });
```

And I think you're missing an `if (!done)` check or something.

Reply:
:-(